apollo-router/src/services/layers/query_analysis.rs

```diff
@@ -257,25 +255,63 @@ pub(crate) struct ParsedDocumentInner {
     pub(crate) ast: ast::Document,
     pub(crate) executable: Arc<Valid<ExecutableDocument>>,
     pub(crate) hash: Arc<QueryHash>,
+    pub(crate) operation: Node<Operation>,
+    /// `__typename`
+    pub(crate) has_root_typename: bool,
+    /// `__schema` or `__type`
+    pub(crate) has_schema_introspection: bool,
+    /// Non-meta fields explicitly defined in the schema
+    pub(crate) has_explicit_root_fields: bool,
 }
 
+#[derive(Debug)]
+pub(crate) struct RootFieldKinds {}
+
 impl ParsedDocumentInner {
-    pub(crate) fn get_operation(
-        &self,
+    pub(crate) fn new(
+        ast: ast::Document,
+        executable: Arc<Valid<ExecutableDocument>>,
         operation_name: Option<&str>,
-    ) -> Result<&Node<Operation>, SpecError> {
-        if let Ok(operation) = self.executable.operations.get(operation_name) {
-            Ok(operation)
-        } else if let Some(name) = operation_name {
-            Err(SpecError::UnknownOperation(name.to_owned()))
-        } else if self.executable.operations.is_empty() {
-            // Maybe not reachable?
-            // A valid document is non-empty and has no unused fragments
-            Err(SpecError::NoOperation)
-        } else {
-            debug_assert!(self.executable.operations.len() > 1);
-            Err(SpecError::MultipleOperationWithoutOperationName)
+        hash: Arc<QueryHash>,
+    ) -> Result<Arc<Self>, SpecError> {
+        let operation = get_operation(&executable, operation_name)?;
+        let mut has_root_typename = false;
+        let mut has_schema_introspection = false;
+        let mut has_explicit_root_fields = false;
+        for field in operation.root_fields(&executable) {
+            match field.name.as_str() {
+                "__typename" => has_root_typename = true,
+                "__schema" | "__type" if operation.is_query() => has_schema_introspection = true,
+                _ => has_explicit_root_fields = true,
+            }
         }
+        Ok(Arc::new(Self {
+            ast,
+            executable,
+            hash,
+            operation,
+            has_root_typename,
+            has_schema_introspection,
+            has_explicit_root_fields,
+        }))
+    }
+}
+
+pub(crate) fn get_operation(
+    executable: &ExecutableDocument,
+    operation_name: Option<&str>,
+) -> Result<Node<Operation>, SpecError> {
+    if let Ok(operation) = executable.operations.get(operation_name) {
+        Ok(operation.clone())
+    } else if let Some(name) = operation_name {
+        Err(SpecError::UnknownOperation(name.to_owned()))
+    } else if executable.operations.is_empty() {
+        // Maybe not reachable?
+        // A valid document is non-empty and has no unused fragments
+        Err(SpecError::NoOperation)
+    } else {
+        debug_assert!(executable.operations.len() > 1);
+        Err(SpecError::MultipleOperationWithoutOperationName)
```

Comment:
This should also be unreachable for a valid document

Reply:
Is it really? `operationName` is not an input of validation, only of [`GetOperation`](https://spec.graphql.org/draft/#GetOperation()) which is exactly this function

Reply:
Ohh yes I see, I misunderstood the condition for this error. There are multiple operations (in the document) without an operation name (in the graphql request).